### PR TITLE
Don't ignore the case when dataflow doesn't know what's going on

### LINKED ISSUE #101026: [NativeAOT] Trivial use of Delegate.Method is not enabling delegate targets for reflection
## Repro

```csharp
class Program
{
    static void Main() => Console.WriteLine(new Action(Test<string>).Method);

    static void Test<T>() { }
}
```

## Actual behavior (current main)

```
Unhandled exception. System.NotSupportedException: Cannot retrieve a MethodInfo for this delegate because the method it targeted (Program.Test[T]()) was not enabled for metadata.
   at Internal.Reflection.Extensions.NonPortable.DelegateMethodInfoRetriever.GetDelegateMethodInfo(Delegate) + 0x276
   at Program.Main() + 0x26
```

## Expected behavior (.NET 8)
```
Void Test[String]()
```

### PR DESCRIPTION
**Pull request by MichalStrehovsky**
Fixes #101026.

TIL that when dataflow doesn't know something, it will model it as no value at all.

### CODE REVIEW

**Comment by MichalStrehovsky** [comment]
I assume it also fixes #101010.

**Comment by agocke** [comment]
Can we unit test this?

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/ReflectionMethodBodyScanner.cs [hunk 1]
```diff
@@ -656,6 +656,12 @@ public override bool HandleCall(MethodIL callingMethodBody, MethodDesc calledMet
                         if (Intrinsics.GetIntrinsicIdForMethod(callingMethodDefinition) == IntrinsicId.RuntimeReflectionExtensions_GetMethodInfo)
                             break;
 
+                        if (param.IsEmpty())
+                        {
+                            // The static value is unknown and the below `foreach` won't execute
```
Comment (jkotas) [inline src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/ReflectionMethodBodyScanner.cs:661]: Why is the value unknown in my repro?

I would expect that the dataflow should be able to trace through `new Action(Test<string>).Method` without problems.

**Comment by MichalStrehovsky** [comment]
> Can we unit test this?

The test would be testing a scenario that only exists due to a dataflow issue we keep running into. Sven just ran into it last week in #100786. Once we fix the underlying issue, we'd have to delete the test because this will no longer be a problem.

We also don't have facilities to unit test this. This would have to be an E2E test and because it's testing "specific pattern inhibits a global optimization in the whole program", it would have to be a brand new test project because this can't be folded into any of the existing ones. All in all, felt like too much hassle to write a test with no longer term benefit.

We do have tests for "global optimization is active" and "global optimization is disabled", just not for this specific pattern that disables it. I'm also a bit confused why this pattern exists in the first place and why dataflow has a distinction between "no known value (`IsEmpty`)" and "an unknown value (`IsUnknown`)".

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/ReflectionMethodBodyScanner.cs (hunk 1 above)
Comment (MichalStrehovsky) [reply src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/ReflectionMethodBodyScanner.cs:661]: Known limitation of the dataflow analysis (#93720 is the tracking issue).

**Comment by am11** [comment]
> I assume it also fixes #101010.

Does it affect the 50 kB save from https://github.com/dotnet/runtime/pull/100916?

**Comment by MichalStrehovsky** [comment]
> > I assume it also fixes #101010.
> 
> Does it affect the 50 kB save from #100916?

No, this doesn't bring back the LINQ expressions that were deleted there.

File: src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/ReflectionMethodBodyScanner.cs (hunk 1 above)
Comment (MichalStrehovsky) [reply src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/ReflectionMethodBodyScanner.cs:661]: > I thought #93720 was about not tracking variable types for `GetType` - how does that explain this issue? If we don't trace through `new Action(Test<string>)`, I would have expected it to show up as an unknown value, not empty.

I thought unknown/empty are interchangeable but now I understand the difference. We eventually want to know the type allocated through the `new Action` do that `GetType` (on in this case Delegate.get_Method`) know the type it's operating on.

In this case, we end up with empty instead of unknown because constructors take this path (they "return void"):

https://github.com/dotnet/runtime/blob/3d0da2cc430333160de47b8345609c75830953ad/src/tools/illink/src/ILLink.Shared/TrimAnalysis/HandleCallAction.cs#L1168-L1170

And we can no longer bash it to unknown here because the HandleCallAction returned true (i.e. `handledFunction`):

https://github.com/dotnet/runtime/blob/3d0da2cc430333160de47b8345609c75830953ad/src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/MethodBodyScanner.cs#L1302-L1319

Comment (MichalStrehovsky) [reply src/coreclr/tools/aot/ILCompiler.Compiler/Compiler/Dataflow/ReflectionMethodBodyScanner.cs:661]: The compiler keeps track of delegate type + method. If we know delegate type, we know all the methods it can be used with.

We don't track this at method granularity because doing `new Action(Foo).Method` is very rare. People only do this if they want to work around lack of `methodof` in C#. What happens 95% of time is that we construct a delegate somewhere and a different part of the program will call `Delegate.get_Method` on it. That's why we only care about delegate type.

Dataflow losing track of the type in the trivial `new Action(Foo).Method` pattern is throwing a wrench in this because if someone uses this to work around `methodof`, we only see `Delegate.get_Method` was called on some unknown thing and we need to disable the optimization globally because this could be any delegate.

**Comment by eerhardt** [comment]
Can we add a test for this? Looks like it broke in dotnet/aspnetcore. https://github.com/dotnet/aspnetcore/pull/55010 broke because of this. It might be good to have a "Native AOT" test that uses DI like this test in dotnet/aspnetcore:

https://github.com/dotnet/aspnetcore/blob/d594fc22076db3616a250817c3ca550fcbe69562/src/DefaultBuilder/test/Microsoft.AspNetCore.NativeAotTests/UseStartupThrowsForStructContainersTest.cs#L20

**Comment by MichalStrehovsky** [comment]
> Can we add a test for this? Looks like it broke in dotnet/aspnetcore. [dotnet/aspnetcore#55010](https://github.com/dotnet/aspnetcore/pull/55010) broke because of this. It might be good to have a "Native AOT" test that uses DI like this test in dotnet/aspnetcore:
> 
> https://github.com/dotnet/aspnetcore/blob/d594fc22076db3616a250817c3ca550fcbe69562/src/DefaultBuilder/test/Microsoft.AspNetCore.NativeAotTests/UseStartupThrowsForStructContainersTest.cs#L20

I already wrote my opinion on testing this in https://github.com/dotnet/runtime/pull/101031#issuecomment-2055314742.

We're running into two issues in dataflow analysis (modeling something that should be Unknown as Empty, and not being able to see through `new Action().Method`). Either of them not existing would be enough to have all codepaths covered by our existing testing. Creating a new test to cover a scenario that is going to be impossible to exercise doesn't seem like a great use of time (and CI resources in the future, since this must be a standalone test that cannot have anything else in it).

Given this is now blocking codeflow we either need to approve and merge this ASAP, or do a revert of the previous PR. @sbomer does this look good? I'm off to bed.

**Review by sbomer — APPROVED** [review]
I'm good with this change for now to unblock dependency flow. I'd like to fix the issue where this is not tracked as "unknown", but I can look into that as a follow-up.

**Comment by eerhardt** [comment]
My intention on testing was more of the end-to-end scenario that broke. So we wouldn't "have to delete the test because this will no longer be a problem.".

> We also don't have facilities to unit test this.

We do have the facilities to test the scenario though. See

https://github.com/dotnet/runtime/blob/69062fde5d234974f16d785174473a1d02e77b49/src/libraries/System.Diagnostics.DiagnosticSource/tests/NativeAotTests/System.Diagnostics.DiagnosticSource.NativeAotTests.proj#L5-L6

https://github.com/dotnet/runtime/blob/69062fde5d234974f16d785174473a1d02e77b49/src/libraries/System.Diagnostics.DiagnosticSource/tests/NativeAotTests/DiagnosticSourceEventSourceTests.cs#L45-L49

We can write a Native AOT app just like the above for the ActivatorUtilities class that was broken in https://github.com/dotnet/runtime/issues/101010. We can basically just use the repro provided in the bug as the test app - all of the referenced code is built in dotnet/runtime. That way we can get a little more "end-to-end" tests here.

**Comment by jkotas** [comment]
Merging to unblock codeflow

**Comment by agocke** [comment]
As general policy, I think every regression should have a regression test. Obviously if it's difficult or impossible to construct such a test, never mind. But if it's straightforward, there's by definition at least one incidence of the bug actually making into the product, which is enough in my opinion to merit the test. We run many thousands of tests on every PR and many of those have never, and will never, catch a bug in production. I'd rather invest in infrastructure improvements in test execution rather than avoid writing tests. We also accidentally revert PRs on occasion and even though we sometimes revert the tests as well, any extra coverage to avoid that is very useful at scale.

**Comment by MichalStrehovsky** [comment]
TL;DR: Any test I can think of writing for this will stop exercising the codepath added here by the end of 9.0, but will forever cost us seconds in CI time because it will have to be the most expensive test kind we have.

You'll have a hard time finding a PR of mine that wouldn't have a test. I do write test for pretty much any code I write. Writing a _good and long term useful test_ exercising the code path I'm adding here is hard, if not impossible. I actually don't know if this codepath is going to be hittable at all once Sven fixes the Empty/Unknown confusion in dataflow; we might want to just revert this PR then.

> We can write a Native AOT app just like the above for the ActivatorUtilities class that was broken in https://github.com/dotnet/runtime/issues/101010. We can basically just use the repro provided in the bug as the test app

This test would not be long term useful. The line that is throwing the exception initializes a field in the cctor that is not even used in native AOT - we set it but never read it. Ironically, the only observable effect of the line is that it needs to disable whole program analysis of delegates due to dataflow analysis deficiencies. It is not the only reason why ASP.NET gets delegate whole program analysis disabled (the other reason is #96528) but once/if it becomes the only reason to disable the optimization in ASP.NET, I'm going to submit another PR that takes it out of statically reachable code on native AOT (I considered doing that in #100916 already and wrote it as an option in PR description). One way or the other, the problematic line is not going to trigger the `if` check I'm adding in this PR by the end of 9.0 (either because the Delegate.get_Method call will be removed from closure, or dataflow will get the necessary improvement).

> We run many thousands of tests on every PR and many of those have never, and will never, catch a bug in production

We spend massive amount of time as a team over many years limiting the number of tests we run. We audited and deleted redundant tests between src/tests and src/libraries. We spent (and are spending) tons of time limiting the number of standalone tests in src/tests.

This can only be tested by adding another fully standalone test (the most expensive test kind we have) that takes seconds to build and will become redundant the moment we fix _any_ of these:

1. Dataflow bug that confuses Unknown with Empty (#101102)
2. Improve tracking of `new` in dataflow (#93720)

1 sounds like a 9.0 bug. 2 would be very nice to fix in 9.0 because we (and our customers - this is customer-reported) keep running into it.

**Comment by martincostello** [comment]
End-to-end tests don't ever really become redundant unless you delete the public APIs they depend on, because they test a whole bunch of things for user scenarios without depending on implementation details. A highly-targeted "unit" test? Sure, that can become redundant as refactoring happens.

The repro is an example of user code (disregard that the delegate did nothing, that was to create a minimal repro to make the specific issue easy to diagnose and fix) that should work regardless of what implementation details it hits or what internals become redundant over time.

If it were made a bit less minimal (say, the delegate set the base address for the requests, as it did in the original app it's derived from) and checked a few extra cases then you'd get more bang-for-your-buck out of it for the time spent compiling it. In the extreme case you could even compile (most of) the whole app I found #101010 in originally and cover even more user scenarios with that one compilation.

**Comment by am11** [comment]
> No, this doesn't bring back the LINQ expressions that were deleted there.

I meant overall stats before the previous PR and after this PR; now we are keeping more symbols Sounds like still a win, was just curious if we re-measured "This makes ASP.NET 50 kB smaller with native AOT".

> another fully standalone test (the most expensive test kind we have)

There are end-to-end tests for AOT which run in PRs of SDK / installer repos. Should we lower the bar of adding tests in those repos which were deemed expensive-for-runtime? It slipped though the cracks and made it to a preview release due to a test gap. Murphy's law says it can happen again.

**Comment by MichalStrehovsky** [comment]
> If it were made a bit less minimal (say, the delegate set the base address for the requests, as it did in the original app it's derived from) and checked a few extra cases then you'd get more bang-for-your-buck out of it for the time spent compiling it. 

The test needs to test **exactly this code path and nothing else**. To hit this bug, the program needs to have a call to `Delegate.get_Method` that hits the bug and no other call to `Delegate.get_Method` that could paper over the bug. Dragging in more code just increases the chances the bug will be papered over. If we try to make this test "more useful", it will just stop being a regression test for this bug: it will become yet another DI test in this repo that cannot catch this kind of bug (we already have a lot of DI tests in this repo that do not catch this bug). For example the repro in https://github.com/martincostello/dotnet-runtime-101010-repro can be fixed by adding `/p:EventSourceSupport=true` to the `dotnet publish` line (EventSourceSupport brings in code that papers over the bug).

On the other hand if we keep the test small and razor sharp testing this, it will stop testing the bug by the end of 9.0 (it's not a question of "if", only "when") and it's questionable what other residual value it will have due to it being focused on one thing only.

The other E2E tests we have in this repo are testing trimming warning suppressions and they will be useful as long as the suppressions exist.

> I meant overall stats before the previous PR and after this PR; now we are keeping more symbols Sounds like still a win, was just curious if we re-measured "This makes ASP.NET 50 kB smaller with native AOT".

ASP.NET doesn't benefit from this optimization yet due to lack of #96528 so I expect the impact of this PR to be zero. (ASP.NET sets EventSourceSupport=true that will paper over this bug because it brings more reasons to disable the optimization.)

>  It slipped though the cracks and made it to a preview release due to a test gap.

Did it? It was hit as ASP.NET core repo was ingesting the runtime build with this.

**Comment by eerhardt** [comment]
> TL;DR: Any test I can think of writing for this will stop exercising the codepath added here by the end of 9.0, but will forever cost us seconds in CI time because it will have to be the most expensive test kind we have.

> It was hit as ASP.NET core repo was ingesting the runtime build with this.

Note that the test in the ASP.NET Core repo that caught this bug wasn't for this exact codepath, but still caught the bug anyway. The test that caught it is named `UseStartupThrowsForStructContainersTest`. It was testing a specific exceptional case in ASP.NET Core, but since it was an end-to-end functional test, it covers a lot of ground.

Adding a new end-to-end functional test for the scenario that broke here will be long term useful because it will ensure the scenario keeps working as we make changes in both in the nativeAOT runtime and in the ActivatorUtilities class.

Not all tests need to be unit tests. Not all tests need to be end-to-end functional tests. We can use both to get the best coverage. Yes, these end-to-end functional tests are expensive - so let's not write hundreds of them. But a handful of them, especially when the scenario they cover is critical to the .NET stack (and has been broken in the past), is important to ensure our changes don't introduce breaks.

Even if the codepath being fixed here will be deleted by the end of 9.0, the scenario that broke isn't going anywhere. Knowing that we will change the codepath makes it even more valuable to test the scenario. The test will ensure we don't break this scenario again when we do make changes.

**Comment by jkotas** [comment]
We have tiered validation system. We do accept that there are going to be bugs that won't get caught by dotnet/runtime pre-checkin tests and that will be only caught by dependent repos or outer loop validation.

We do have https://github.com/dotnet/runtime/tree/main/src/libraries/Microsoft.Extensions.DependencyInjection/tests/TrimmingTests that is basic end-to-end test that exercises DI and ActivatorUtilities in a simple app. Does this test run in native AOT outer loop runs? If not, it is something to fix.

The question to ask before adding new end-to-end tests: How many breaks would have been caught by the new end-to-end tests in last year?

There is dotnet/runtime bug caught by dependent repos every other week. It is a different end-to-end scenarios and different corner case each time. We cannot afford to add a new end-to-end test for each of these.

**Comment by eerhardt** [comment]
> We do have https://github.com/dotnet/runtime/tree/main/src/libraries/Microsoft.Extensions.DependencyInjection/tests/TrimmingTests that is basic end-to-end test that exercises DI and ActivatorUtilities in a simple app. Does this test run in native AOT outer loop runs? If not, it is something to fix.

No, those tests only run with `PublishTrimmed=true`. They don't run with `PublishAot=true`. You can tell becuase the project is named `TrimmingTests.proj`. See

https://github.com/dotnet/runtime/blob/6f3b1a6784818684622ac32a910b1f333708ea82/Directory.Build.props#L399-L400

**Comment by am11** [comment]
> Did it? It was hit as ASP.NET core repo was ingesting the runtime build with this.

https://github.com/dotnet/runtime/issues/101026#issue-2242266245 repros in 9.0 p3 and does not repros with 8.0. So no test in any repo has caught this regression in a while.

**Comment by am11** [comment]
> Did it? It was hit as ASP.NET core repo was ingesting the runtime build with this.

https://github.com/dotnet/runtime/issues/101026#issue-2242266245 repros in 9.0 p3 and does not repro with 8.0. So no test in any repo has caught this regression in a while.

**Comment by MichalStrehovsky** [comment]
> > Did it? It was hit as ASP.NET core repo was ingesting the runtime build with this.
> 
> [#101026 (comment)](https://github.com/dotnet/runtime/issues/101026#issue-2242266245) repros in 9.0 p3 and does not repro with 8.0. So no test in any repo has caught this regression in a while.

For the purposes of **delegate optimization** that I'm changing here, this:

```csharp
class Program
{
    static void Main() => Console.WriteLine(new Action(Test<string>).Method);

    static void Test<T>() { }
}
```

Is no different from this:

```csharp
class Program
{
    static Action GetTest() => new Action(Test<string>);

    static void Main() => Console.WriteLine(GetTest().Method);

    static void Test<T>() { }
}
```

Or this:

```csharp
class Program
{
    static Action TheTest => new Action(Test<string>);

    static void Main() => Console.WriteLine(TheTest.Method);

    static void Test<T>() { }
}
```

The latter two do not run into this bug. I could come up with many more that do not run into the bug. The delegate optimization does not concern itself with these shapes. Dataflow analysis does. We have delegate optimization tests that tests _one_ of these shapes.

We're **not** going to add E2E standalone tests for all of these variations that take seconds to build to test all possible variations. It is the job of the dataflow analysis to test these variations in it's unit tests that take milliseconds to run and are the place where dataflow analysis is tested.

*I'm not fixing the dataflow analysis bug, merely working around it*. The fix in the data flow analysis is tracked in #101102. Once that's fixed, my workaround in this PR won't even be needed, it will just cover some corner case I don't even know how to hit (I'll leave it to Sven whether we should just roll back this PR then).

We cannot unit test the delegate optimization. We can unit test dataflow. We do not test dataflow with delegate optimization tests. We do not test dataflow analysis with DI tests. We test it with dataflow tests.
